### PR TITLE
[VOL-79] Refactor internal API

### PR DESCRIPTION
**Pull request by lukemarsden**
Introduces a layer of indirection which allows the CLI to talk "dvol language" to an internal dvol API, and the internal dvol API to speak "data layer language" to the Data Layer implementation. Also introduces types and methods for great win.

This was initially the start of the checkout branch, but all checkout-related functionality has been removed.

This is tidy-up in advance of introducing more commit and branch related functionality in `checkout` and `branch`.

Fixes https://clusterhq.atlassian.net/browse/VOL-79

### CODE REVIEW

File: pkg/cmd/utils.go [hunk 1]
```diff
@@ -15,6 +15,16 @@ func checkVolumeArgs(args []string) error {
 	return nil
 }
 
+func checkBranchArgs(args []string) error {
```
Comment (zubron) [inline pkg/cmd/utils.go]: This function isn't being used elsewhere as part of this change.

File: pkg/cmd/rm_test.go [hunk 2]
```diff
@@ -34,7 +34,7 @@ func TestRmInvalidVolumeName(t *testing.T) {
 
 func TestRmVolumeDoesNotExist(t *testing.T) {
 	// Setup
-	originalBasePath := basePath
+	originalBasepath := basePath
```
Comment (zubron) [inline pkg/cmd/rm_test.go]: :question: Can we revert this name change since it isn't needed?

File: pkg/datalayer/datalayer.go [hunk 3]
```diff
@@ -1,113 +1,33 @@
...
-	volumePath := filepath.FromSlash(basePath + "/" + volumeName)
+func (dl *DataLayer) CreateVolume(volumeName string) error {
+	volumePath := filepath.FromSlash(dl.BasePath + "/" + volumeName)
 	// TODO Factor this into a data layer object.
 	err := os.MkdirAll(volumePath, 0777) // XXX SEC
 	if err != nil {
 		return err
 	}
-	return setActiveVolume(basePath, volumeName)
+	return nil
```
Comment (zubron) [inline pkg/datalayer/datalayer.go]: :question: This function could now just return `os.MkdirAll..`

File: pkg/datalayer/datalayer.go [hunk 4]
```diff
@@ -1,113 +1,33 @@
...
-	return nil
+type DataLayer struct {
+	BasePath string
 }
 
-func CreateVolume(basePath string, volumeName string) error {
-	volumePath := filepath.FromSlash(basePath + "/" + volumeName)
+func (dl *DataLayer) CreateVolume(volumeName string) error {
+	volumePath := filepath.FromSlash(dl.BasePath + "/" + volumeName)
 	// TODO Factor this into a data layer object.
```
Comment (zubron) [inline pkg/datalayer/datalayer.go]: :question: Is this comment still needed?

**Comment by zubron** [comment]
Looks good, thanks @lukemarsden and @brendan-donegan! Address the comments and then merge when the tests pass.

**Comment by lukemarsden** [comment]
Thanks! Addressed review feedback, now waiting for tests, then will merge if green.

**Comment by zubron** [comment]
Tests are green, I'm going to merge :smile:

**Comment by lukemarsden** [comment]
Thanks, @zubron!

**Pull request merged**